SparseGrids/tsgCoreOneDimensional.cpp

```diff
@@ -392,6 +393,168 @@ const char* OneDimensionalMeta::getHumanString(TypeOneDRule rule){
+        for (int k=0; k<i; k++) {
+            offdiag[k] = sqrt(beta[k]);
+        }
+        TasmanianTridiagonalSolver::getSymmetricEigenvalues(i+1, roots, offdiag);
+        if (roots.size() % 2 == 1) {
+            roots[(roots.size() - 1) / 2] = 0.0; // Zero out the center for stability.
+        }
+        root_cache[i] = roots;
+        // Update the values of the polynomials at ref_points.
+        poly_m1_vals = poly_vals;
+        std::transform(ref_points.begin(), ref_points.end(), poly_vals.begin(),
+                       [&roots, i](double x){return poly_eval(roots, x);});
+    }
+    #endif
+    return root_cache;
+}
+
+// Get Exotic Gauss-Legendre points and weights. For the case where the
+// integrand F(x) is of the form:
+//     F(x) := f(x) * [weight_fn(x) - shift] + shift * f(x)
+double lagrange_eval(int idx, const std::vector<double> roots, double x) {
+    double eval_value = 1.0;
+    for (size_t i=0; i<roots.size(); i++) {
+        if (i != idx) {
+            eval_value *= (x - roots[i]) / (roots[idx] - roots[i]);
+        }
+    }
+    return eval_value;
+}
+void OneDimensionalExoticQuad::getExoticGaussLegendreCache(
+        std::vector<std::vector<double>> &weights_cache,
+        std::vector<std::vector<double>> &points_cache,
+        const int n,
+        const double shift,
+        std::function<double(double)> weight_fn,
+        const int nref) {
+
+    // Create the set of points for the first term.
+    std::vector<double> ref_weights(nref), ref_points(nref);
+    TasGrid::OneDimensionalNodes::getGaussLegendre(nref, ref_weights, ref_points);
+    points_cache.resize(n);
+    points_cache = OneDimensionalExoticQuad::getRootCache(
+        n, [shift, weight_fn](double x)->double{return (weight_fn(x) + shift);},
+        ref_weights, ref_points);
+
+    // Create the set of weights for the first term.
+    weights_cache.resize(n);
+    for (int i=0; i<n; i++) {
+        std::vector<double> quad_weights(points_cache[i].size(), 0.0);
+        for (size_t j=0; j<points_cache[i].size(); j++) {
+            // Integrate the function l_j(x) * [weight_fn(x) + shift].
+            for (int k=0; k<nref; k++) {
+                quad_weights[j] +=
+                        lagrange_eval(j, points_cache[i], ref_points[k]) *
+                        (weight_fn(ref_points[k]) + shift) *
+                        ref_weights[k];
+            }
+        }
+        weights_cache[i] = quad_weights;
+    }
+
+    // Create and append the set of points and weights for the second term.
+    for (int i=0; i<n; i++) {
+        std::vector<double>
+                points(points_cache[i].size()),
+                gl_weights(points_cache[i].size());
+        TasGrid::OneDimensionalNodes::getGaussLegendre(points_cache[i].size(),
+                                                       gl_weights,
+                                                       points);
+        std::vector<double> quad_weights(points_cache[i].size());
```

Comment:
No need for quad_weights, you can just modify gl_weights

Reply:
Changed.